### PR TITLE
feature/433 a learner can book onto a training course m

### PR DESCRIPTION
**Pull request by superafroman**
All of the good bookings stuff.

### CODE REVIEW

File: src/ui/controllers/booking.ts [hunk 1]
```diff
@@ -181,19 +195,25 @@ export async function tryCancelBooking(
 	res: express.Response
 ) {
 	const course = req.course
+	const record = await learnerRecord.getCourseRecord(req.user, course)
+
+	if (!record.selectedDate) {
+		res.redirect('/')
```
Comment (tav) [inline src/ui/controllers/booking.ts]: Is this where we want to redirect them?

File: src/lib/youtube.ts [hunk 2]
```diff
@@ -11,32 +11,8 @@ export interface BasicInfo {
...
-	duration += (parseInt(parts[7], 10) || 0) * 60
-	duration += (parseInt(parts[6], 10) || 0) * 3600
-	duration += (parseInt(parts[4], 10) || 0) * 86400
-	// Accept the leading minus sign for now, but might want to abort in future.
-	if (parts[1]) {
-		return -duration
-	}
-	return duration
+function convertDuration(isoDuration: string): string {
+	return isoDuration.substr(2).toLowerCase()
```
Comment (tav) [inline src/lib/youtube.ts]: Is this format good enough?

File: src/ui/component/Learning.html [hunk 3]
```diff
@@ -11,13 +11,17 @@
                 </a>
             {{/if}}
             {{#if course.state}}
-                {{#if course.type === 'elearning'}}
-                <a href="/courses/{{course.uid}}/do">Resume</a>
+                {{#if course.type === 'face-to-face'}}
```
Comment (tav) [inline src/ui/component/Learning.html]: Are we not calling this `event` like yuksel proposed?

Comment (superafroman) [reply src/ui/component/Learning.html]: The course is a face to face course which has events - had a follow on discussion around this with the business last week.

File: src/lib/youtube.ts (hunk 2 above)
Comment (superafroman) [reply src/lib/youtube.ts]: Yeah, it's consistent with how they label the other course durations.

File: src/ui/controllers/booking.ts (hunk 1 above)
Comment (superafroman) [reply src/ui/controllers/booking.ts]: Ideally we'd send them some malware and take over their computer, but I don't think I'd get away with that. A user will only hit this if they're screwing around trying to do things they shouldn't so I'm not too bothered about the experience at the mo.

File: src/ui/component/Learning.html (hunk 3 above)
Comment (tav) [reply src/ui/component/Learning.html]: Aha!

**Pull request merged**